libp2p.go

```diff
@@ -0,0 +1,102 @@
+package libp2p
+
+import (
+	"context"
+	"crypto/rand"
+
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	host "github.com/libp2p/go-libp2p-host"
+	pnet "github.com/libp2p/go-libp2p-interface-pnet"
+	metrics "github.com/libp2p/go-libp2p-metrics"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	transport "github.com/libp2p/go-libp2p-transport"
+	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+	mux "github.com/libp2p/go-stream-muxer"
+	ma "github.com/multiformats/go-multiaddr"
+	mplex "github.com/whyrusleeping/go-smux-multiplex"
+	msmux "github.com/whyrusleeping/go-smux-multistream"
+	yamux "github.com/whyrusleeping/go-smux-yamux"
+)
+
+// Config describes a set of settings for a libp2p node
+type Config struct {
+	Transports  []transport.Transport
+	Muxer       mux.Transport
+	ListenAddrs []ma.Multiaddr
+	PeerKey     crypto.PrivKey
+	Peerstore   pstore.Peerstore
+	Protector   pnet.Protector
+	Reporter    metrics.Reporter
+}
+
+func New(ctx context.Context) (host.Host, error) {
+	return NewWithCfg(ctx, DefaultConfig())
+}
+
+// Construct instantiates a libp2p host using information from the given
+// config. `nil` may be passed to use default options.
+func NewWithCfg(ctx context.Context, cfg *Config) (host.Host, error) {
+	if cfg == nil {
+		cfg = DefaultConfig()
+	}
+
+	// If no key was given, generate a random 2048 bit RSA key
+	if cfg.PeerKey == nil {
+		priv, _, err := crypto.GenerateKeyPairWithReader(crypto.RSA, 2048, rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		cfg.PeerKey = priv
+	}
+
+	// Obtain Peer ID from public key
+	pid, err := peer.IDFromPublicKey(cfg.PeerKey.GetPublic())
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new blank peerstore if none was passed in
+	ps := cfg.Peerstore
+	if ps == nil {
+		ps = pstore.NewPeerstore()
+	}
+
+	ps.AddPrivKey(pid, cfg.PeerKey)
+	ps.AddPubKey(pid, cfg.PeerKey.GetPublic())
+
+	swrm, err := swarm.NewSwarmWithProtector(ctx, cfg.ListenAddrs, pid, ps, cfg.Protector, cfg.Muxer, cfg.Reporter)
+	if err != nil {
+		return nil, err
+	}
+
+	netw := (*swarm.Network)(swrm)
+
+	return bhost.New(netw), nil
+}
+
+func DefaultMuxer() mux.Transport {
+	// Set up stream multiplexer
+	tpt := msmux.NewBlankTransport()
+
+	// By default, support yamux and multiplex
```

Comment:
It strikes me that here the supported transports are slightly different than here: https://github.com/libp2p/go-libp2p-swarm/blob/master/swarm.go

I thought that `swarm.NewNetworkWithProtector` would be the *default* way of doing things so any updates to that should happen in `swarm`?

At first sight, it seems that this function is just re-doing what PSTransport provides:

```
// PSTransport is the default peerstream transport that will be used by
// any libp2p swarms.
var PSTransport pst.Transport
```

Reply:
Good catch, I'm actually thinking the best path forward here is to remove that logic from the swarm code, and make this place be the default. The reason being that swarm is much deeper in the dependency tree, and making it depend on extra packages makes it harder to update those packages.

Reply:
That sounds good to me 👍